### ISSUE TITLE
Build error: missing modules on Ubuntu

### ISSUE THREAD

**Issue by johncarlosignacio** [jasp-stats/jasp-issues#655]
* JASP version: git clone
* OS name and version: Ubuntu 19.10
* Analysis: Trying to build from source but Makefile is not found during qmake (since Flatpack installation fails with service unavailable)
* Bug description: Makefile (size is 34.3 kb) is in directory, but error msg returned on qmake: `Could not find absolute location of file  "make: *** [/home/carli/Softwares/build-JASP-Desktop-Debug/Makefile"`
* Expected behaviour: Proceed with qmake

#### Steps to reproduce:
1. Cloned repository
2. `$ sudo qtcreator JASP.pro`
3. Run qmake

#### Terminal log:
```
$ sudo qtcreator JASP.pro
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
Info: creating stash file /home/carli/Softwares/jasp-desktop/.qmake.stash
Info: creating stash file /home/carli/Softwares/jasp-desktop/JASP-Common/.qmake.stash
Info: creating stash file /home/carli/Softwares/jasp-desktop/JASP-R-Interface/.qmake.stash
Info: creating stash file /home/carli/Softwares/jasp-desktop/JASP-Engine/.qmake.stash
Info: creating stash file /home/carli/Softwares/jasp-desktop/JASP-Desktop/.qmake.stash
Info: creating stash file /home/carli/Softwares/build-JASP-Desktop-Debug/.qmake.stash
Info: creating stash file /home/carli/Softwares/build-JASP-Desktop-Debug/JASP-Desktop/.qmake.stash
Info: creating stash file /home/carli/Softwares/build-JASP-Desktop-Debug/JASP-Common/.qmake.stash
Info: creating stash file /home/carli/Softwares/build-JASP-Desktop-Debug/JASP-R-Interface/.qmake.stash
Info: creating stash file /home/carli/Softwares/build-JASP-Desktop-Debug/JASP-Engine/.qmake.stash
Could not find absolute location of file  "make: *** [/home/carli/Softwares/build-JASP-Desktop-Debug/Makefile"
```

**Comment by JorisGoosen** [comment]
First of: you really shouldn't be running qtcreator as root user, you can just start qtcreator normally.
Secondly: Did you follow [the building guide for linux](https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/jasp-building-guide.md#linux)

And last but certainly not least: maybe you should give flatpak another try?
- Possibly the server was down when you tried, trying it again might help.
- Otherwise it could be the case that you are running an older version of flatpak, Ubuntu has the tendency to be slow in updating that.You can check by running `flatpak --version`, I have "Flatpak 1.6.2"

If you want to make sure you always have an up to date version of flatpak you can use [this launchpad ppa](https://launchpad.net/~alexlarsson/+archive/ubuntu/flatpak) which is operated by the main author of flatpak.

**Comment by johncarlosignacio** [comment]
I have tried both sudo and non-sudo with same results. Yes, I followed that guide.

I updated flatpak from 1.4.3 to 1.7.1. Still 503: Service Unavailable. I have tried this for the past 3 weeks getting the same results. The download just stops and won't complete. This is the same case when using Ubuntu Software Center (USC). I also noticed in USC that download reaches 100% shortly then the percentage shrinks as the download progresses.

**Comment by JorisGoosen** [comment]
Ok so for troubleshooting flatpak: could you give me the output of `flatpak remotes`?

And what happens if you just start qtcreator and then open JASP.pro as a project through the GUI?

**Comment by johncarlosignacio** [comment]
Running non-sudo `qtcreator JASP.pro`.
![image](https://user-images.githubusercontent.com/6889059/78279825-1e6d4880-74e6-11ea-8088-b574c0946a78.png)

In the general message, I noticed only now that it shows `Project ERROR: Unknown module(s) in QT: quickcontrols2`.
![image](https://user-images.githubusercontent.com/6889059/78280415-0c3fda00-74e7-11ea-9393-c066ecb43ffb.png)

Here is the error running qmake. In terminal, `Could not find absolute location of file  "make: *** [/home/carli/Softwares/build-JASP-Desktop-Debug/Makefile"`.
![image](https://user-images.githubusercontent.com/6889059/78279909-40ff6180-74e6-11ea-8e3d-d4b97afb1b7b.png)

**Comment by JorisGoosen** [comment]
Aah it seems the building guide is missing the fact that you also need to install (at least) `qml-module-qtquick-controls2`.

I guess it wasn't updated for a while as the previous LTS version of ubuntu was so outdated with regard to qt that we sort of stopped looking at it. I would be grateful if you tested out if you can build it after installing that package. Then I can amend the building-guide.

I do think the flatpak workaround should help you avoid this problem entirely though, the "try it until it works" approach seems to solve the problem for other people with the same problem.

**Comment by johncarlosignacio** [comment]
After some iterations, I was able to complete installation but it only downloaded 96 out of 377 MB. 

```
$ until flatpak install flathub org.jaspstats.JASP -y; do echo "Trying again"; sleep 2; done
Looking for matches…

org.jaspstats.JASP permissions:
    ipc                  network              wayland      x11      dri
    file access [1]      dbus access [2]

    [1] home, xdg-config/kdeglobals:ro, xdg-run/dconf, ~/.config/dconf:ro
    [2] com.canonical.AppMenu.Registrar


        ID                          Branch         Op         Remote          Download
 1. [✗] org.jaspstats.JASP          stable         i          flathub         260.3 MB / 377.8 MB

Error: While pulling app/org.jaspstats.JASP/x86_64/stable from remote flathub: Connection terminated unexpectedly
error: Failed to install org.jaspstats.JASP: While pulling app/org.jaspstats.JASP/x86_64/stable from remote flathub: Connection terminated unexpectedly
Trying again
...
[message clipped]
...
Looking for matches…

org.jaspstats.JASP permissions:
    ipc                  network              wayland      x11      dri
    file access [1]      dbus access [2]

    [1] home, xdg-config/kdeglobals:ro, xdg-run/dconf, ~/.config/dconf:ro
    [2] com.canonical.AppMenu.Registrar


        ID                          Branch         Op         Remote          Download
 1. [✓] org.jaspstats.JASP          stable         i          flathub         96.6 MB / 377.8 MB

Installation complete.
```

With `flatpak run org.japstats.JASP`, it tries to run and stalls with a black screen. I am guessing it was not a successful installation.

Installing `qml-module-qtquick-controls2` did not solve the Makefile problem. I really hope I am able to provide a more useful diagnosis but I cannot think of anything.

**Comment by JorisGoosen** [comment]
Too bad that flathub is behaving so badly, I haven't really noticed this problem myself nor heard about it often. But those linked issues seemed to be by people who encounter it more often.. So let's leave that at that for now.

For a more useful diagnosis of your current problem, now that you have installed quick-controls-2 we can see if it is missing anything now. Perhaps you could open qtcreator again and choose "run qmake" and show me the log again? Maybe this time it is complaining about another one.

Besides that, I think @vandenman also builds on ubuntu? Maybe he or @JohnnyDoorn remembers which packages werere needed there and tell us if we are missing something.

It would be nice to bring the building-guide up to date. And otherwise I can also do a clean install of ubuntu tomorrow and try it out myself.

**Comment by johncarlosignacio** [comment]
After installing the missing module, there is no more project error but "run qmake" still produces the same error `Could not find absolute location of file  "make: *** [/home/carli/Softwares/build-JASP-Desktop-Debug/Makefile"`

**Comment by JorisGoosen** [comment]
And can you actually run `make` in a terminal?

Maybe you still need to install `build-essentials`?

**Comment by johncarlosignacio** [comment]
First run of `make` ended in `Project ERROR: Unknown module(s) in QT: quickcontrols2` and found out [here](https://forum.qt.io/topic/90472/unknown-module-s-in-qt-quickcontrols2/5) I had to install `qtquickcontrols2-5-dev`.

Rerun of `make`ran for a while and stopped with error:
```
/usr/bin/ld: cannot find -lreadstat
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1115: ../jasp] Error 1
make[1]: Leaving directory '/home/carli/Softwares/build-JASP-Desktop-Debug/JASP-Desktop'
make: *** [Makefile:101: sub-JASP-Desktop-make_first] Error 2
```

I checked if I forgot to copy `libreadstat.la` but it is there in the build folder.
```
carli@x395:~/Softwares/build-JASP-Desktop-Debug$ pwd
/home/carli/Softwares/build-JASP-Desktop-Debug
carli@x395:~/Softwares/build-JASP-Desktop-Debug$ ls
JASP-Common   JASP-R-Interface          Makefile          writeImage.R
JASP-Desktop  libJASP-Common.a          R                 zzzWrappers.R
JASP-Engine   libJASP-R-Interface9.7.a  Resources
JASPEngine    libreadstat.la            versionScript.sh
```

Installing `build-essential` says I already have the latest version installed.

**Comment by JorisGoosen** [comment]
Ah right, yeah the `-dev` packages.

So it's good you checked to see if you got `libreadstat`, the weird thing though, is that I expected to see `libreadstat.a` and if I check the {github view of `jasp-required-files` I see the same](https://github.com/jasp-stats/jasp-required-files/tree/Linux). Maybe try that one? 
Where did you libreadstat.la from? You build it yourself? In that case you should have the .a variant as well I guess.

**Comment by johncarlosignacio** [comment]
Ahh I see, it was my bad, I copied the wrong `libreadstat` file. I was able to run `make` after copying the right file and it successfully compiled. What do you think is the problem?

I ran the compiled `./jasp` and it produced logs (see below) about packages that were either missing or had a wrong version.

I am trying to install all the right packages now using `installExpectedPackages.R`. I ran into some hiccups when `get-r-package-info.R` was returning packages that had moved to bioconductor already, I had to install them manually. I am still on the process of installing them and I will update you after.

```
carli@x395:~/Softwares/build-JASP-Desktop-Debug$ ./jasp
QML debugging is enabled. Only use this in a safe environment.
Desktop:	Desktop:	JASP 0.12 is initializing.
Desktop:	R_HOME set to /usr/lib/R
Desktop:	JASP Desktop started and Engines initalized.
Desktop:	Initializing QML
Desktop:	Loading Themes
Desktop:	Engine#0:	Log and possible redirects initialized!
Engine#0:	jaspEngine started and has slaveNo 0 and it's parent PID is 11983
Desktop:	Loading HelpWindow
QQmlApplicationEngine failed to load component
qrc:/components/JASP/Widgets/HelpWindow.qml:3 module "QtWebEngine" is not installed

Could not load QML: qrc:/components/JASP/Widgets/HelpWindow.qml
carli@x395:~/Softwares/build-JASP-Desktop-Debug$ [1] ".outputSink initialized!"
Evaluating R-code:
jaspResultsModule$jaspTable$methods(addColumnInfo = function(name=NULL, title=NULL, overtitle=NULL, type=NULL, format=NULL, combine=NULL) { addColumnInfoHelper(name, title, type, format, combine, overtitle) })
Output:

Evaluating R-code:
jaspResultsModule$jaspTable$methods(addFootnote =   function(message='', symbol=NULL, col_names=NULL, row_names=NULL) { addFootnoteHelper(message, symbol, col_names, row_names) })
Output:

Evaluating R-code:
suppressPackageStartupMessages(library("JASP"))
Output:

Evaluating R-code:
suppressPackageStartupMessages(library("JASPgraphs"))
Output:
Error: package or namespace load failed for ‘JASPgraphs’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):
 namespace ‘rlang’ 0.4.1 is already loaded, but >= 0.4.3 is required

Evaluating R-code:
suppressPackageStartupMessages(library("methods"))
Output:

Evaluating R-code:
suppressPackageStartupMessages(library("modules"))
Output:

Evaluating R-code:
source(file='writeImage.R')
Output:

Evaluating R-code:
source(file='zzzWrappers.R')
Output:

Evaluating R-code:
initEnvironment()
Output:

R_HOME: /usr/lib/R/
Evaluating R-code:
.automaticColumnEncDecoding <- FALSE
Output:

Engine#0:	jaspRCPP_init was run and R_HOME: Evaluating R-code:
R.home('')
Output:

/usr/lib/R/
Engine#0:	{
   "messages" : [
      "Package Epi not installed!",
      "Package Hmisc is not the correct version; expected: 4.3-0, installed: 4.4-0",
      "Package IsingSampler is not the correct version; expected: 0.2, installed: 0.2.1",
      "Package MASS is not the correct version; expected: 7.3-51.4, installed: 7.3-51.5",
      "Package Matrix is not the correct version; expected: 1.2-17, installed: 1.2-18",
      "Package NetworkToolbox is not the correct version; expected: 1.3.2, installed: 1.4.0",
      "Package OpenMx not installed!",
      "Package TTR not installed!",
      "Package afex not installed!",
      "Package base is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package boot is not the correct version; expected: 1.3-23, installed: 1.3-24",
      "Package bootnet is not the correct version; expected: 1.2.4, installed: 1.3",
      "Package bridgesampling not installed!",
      "Package broom not installed!",
      "Package bstats is not the correct version; expected: 0.0.0.9001, installed: 1.1-11-5",
      "Package car not installed!",
      "Package cellranger not installed!",
      "Package colourpicker not installed!",
      "Package compiler is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package conting not installed!",
      "Package cowplot not installed!",
      "Package datasets is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package dplyr is not the correct version; expected: 0.8.3, installed: 0.8.5",
      "Package emmeans not installed!",
      "Package flexplot is not the correct version; expected: 0.7.3, installed: 0.7.5",
      "Package forcats is not the correct version; expected: 0.4.0, installed: 0.5.0",
      "Package foreign is not the correct version; expected: 0.8-71, installed: 0.8-75",
      "Package fpc not installed!",
      "Package gdtools not installed!",
      "Package ggdendro not installed!",
      "Package ggm not installed!",
      "Package ggplot2 is not the correct version; expected: 3.2.1, installed: 3.3.0",
      "Package ggrepel not installed!",
      "Package glmnet not installed!",
      "Package grDevices is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package graphics is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package grid is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package haven not installed!",
      "Package htmlTable is not the correct version; expected: 1.13.2, installed: 1.13.3",
      "Package jomo not installed!",
      "Package lavaan not installed!",
      "Package lifecycle is not the correct version; expected: 0.1.0, installed: 0.2.0",
      "Package lmerTest not installed!",
      "Package maptools not installed!",
      "Package markdown not installed!",
      "Package methods is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package mgcv is not the correct version; expected: 1.8-28, installed: 1.8-31",
      "Package mi not installed!",
      "Package mice not installed!",
      "Package mitml not installed!",
      "Package networktools is not the correct version; expected: 1.2.1, installed: 1.2.2",
      "Package nlme is not the correct version; expected: 3.1-140, installed: 3.1-144",
      "Package openxlsx not installed!",
      "Package parallel is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package pbkrtest not installed!",
      "Package pillar is not the correct version; expected: 1.4.2, installed: 1.4.3",
      "Package qgraph is not the correct version; expected: 1.6.3, installed: 1.6.5",
      "Package quantmod not installed!",
      "Package readr not installed!",
      "Package readxl not installed!",
      "Package regsem is not the correct version; expected: 1.3.9, installed: 1.5.2",
      "Package remotes is not the correct version; expected: 2.1.0, installed: 2.1.1",
      "Package rio not installed!",
      "Package rstan is not the correct version; expected: 2.19.2, installed: 2.19.3",
      "Package rstantools not installed!",
      "Package sem not installed!",
      "Package semTools not installed!",
      "Package shiny is not the correct version; expected: 1.4.0, installed: 1.4.0.2",
      "Package shinyjs not installed!",
      "Package smacof is not the correct version; expected: 2.0-0, installed: 2.1-0",
      "Package splines is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package stats is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package stats4 is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package survival is not the correct version; expected: 2.44-1.1, installed: 3.1-8",
      "Package tcltk is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package tibble is not the correct version; expected: 2.1.3, installed: 3.0.0",
      "Package tidyr is not the correct version; expected: 1.0.0, installed: 1.0.2",
      "Package tidyselect is not the correct version; expected: 0.2.5, installed: 1.0.0",
      "Package tools is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package tseries not installed!",
      "Package utils is not the correct version; expected: 3.6.1, installed: 3.6.3",
      "Package weights is not the correct version; expected: 1.0, installed: 1.0.1",
      "Package zip not installed!"
   ],
   "official" : false
}

```

**Comment by johncarlosignacio** [comment]
I tried qmake and now it works fine. I am now sure what fixed it, maybe because of running the Rscripts successfuly without any errors.

**Comment by johncarlosignacio** [comment]
During `Build All`, here is the error for JASP-desktop
```
make[1]: Leaving directory '/home/carli/Softwares/build-JASP-Desktop-Debug/JASP-Desktop'
make: *** [Makefile:101: sub-JASP-Desktop-make_first] Error 2
00:38:16: The process "/usr/bin/make" exited with code 2.
Error while building/deploying project JASP (kit: Desktop)
When executing step "Make"
```

**Comment by JorisGoosen** [comment]
Maybe there were some errors before?

Maybe you can past the entire compile log, because the actual error that caused make to stop the process is probably somewhere earlier. You can also search for "error" but I do not mind helping.

**Comment by JorisGoosen** [comment]
Aah and I see you are also missing `qml-module-qtwebengine`. Maybe a `-dev` variant of this as well, I checked [here](https://packages.ubuntu.com/eoan/qml-module-qtwebengine) but I couldn't see such a variant.

**Comment by johncarlosignacio** [comment]
There were a total of 4 errors from `Build All` using qtcreator. Aside from `qml-module-qtwebengine`, I had to install`qml-module-qtwebchannel` and `qml-module-qt-labs-folderlistmodel`. That left me with only one error which was:

`ERROR: lazy loading failed for package 'JASPgraphs'`

I think the main reason was that I was [using a different version](https://github.com/jasp-stats/jasp-issues/issues/375) of `R-3.6.3`, so I downgraded to `R-3.6.1`. I also installed `JASPgraphs` with `remotes::install_github("vandenman/JASPgraphs")` but I am not sure if that helped at all.

**Finally,** I was able to run `./jasp` successfully.

Thanks for all the help and guidance, Joris!

Here are all the deb modules I installed:

`sudo apt-get install qml-module-qtquick-controls2 qtquickcontrols2-5-dev qml-module-qtwebengine qml-module-qtwebchannel qml-module-qt-labs-folderlistmodel`

As I mentioned earlier, I was not able to successfully run `Rscript installExpectedPackages.R` right of the bat as it resulted to index out of bounds error even with R-3.6.1 (I noticed you committed a fixed for this already). I had to remove the `graph` package by editing `Tools/flatpak/gather-r-package-info.R` since it was not in CRAN anymore. Then installed it with:
```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("graph")
```
Lastly, I tried reproducing my original complaint using a fresh clone and was not able to produce it anymore. I think installing the missing modules fixed the problem.
